components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/cache/BaseCache.java

```diff
@@ -182,6 +182,52 @@ public void addToCache(K key, V entry, int tenantId) {
         }
     }
 
+    /**
+     * Add a cache entry if the same value is not already there.
+     *
+     * @param key   Key which cache entry is indexed.
+     * @param entry Actual object where cache entry is placed.
+     */
+    public void addToCacheIfNoDuplicate(K key, V entry, String tenantDomain) {
+
+        if (!isEnabled()) {
+            return;
+        }
+
+        try {
+            startTenantFlow(tenantDomain);
+            Cache<K, V> cache = getBaseCache();
+            if (cache != null) {
+                cache.putIfNoDuplicate(key, entry);
+            }
+        } finally {
+            PrivilegedCarbonContext.endTenantFlow();
+        }
+    }
+
+    /**
+     * Add a cache entry if the same value is not already there.
+     *
+     * @param key   Key which cache entry is indexed.
+     * @param entry Actual object where cache entry is placed.
+     */
+    public void addToCacheIfNoDuplicate(K key, V entry, int tenantId) {
+
+        if (!isEnabled()) {
+            return;
+        }
+
+        try {
+            startTenantFlow(tenantId);
```

Comment:
Will the startTenantFlow work with both tenantId and tenantDomain?? Please check

Reply:
it does 
https://github.com/wso2/carbon-identity-framework/blob/20aa45abf92629baed29c501f36bf67d9c14b6b1/components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/cache/BaseCache.java#L372